Snapshotting support doc (#1890)
firecracker-microvm/firecracker: docs/snapshotting/snapshot-support.md

in multiple files that are composing the full microVM snapshot:
- the guest memory file,
- the microVM state file,
- zero or more disk files (depending on how many the guest had; these are
  **managed by the users**, which means they need to externally back up their
  block devices backing files).

This design allows sharing any read-only resources (such as read-only disks or
microVM state) between multiple microVMs in order to improve density.
When loading a snapshot, instead of loading at resume time the full contents from
file to memory, Firecracker creates a `MAP_PRIVATE` mapping of the memory file,
resulting in runtime on-demand loading of memory pages. Any subsequent memory writes
go to a copy-on-write anonymous memory mapping.
This has the advantage of very fast snapshot loading times, but comes with the cost
of having to keep the guest memory file around for the entire lifetime of the
resumed microVM.

## Firecracker Snapshotting characteristics

- Fresh Firecracker microVMs are booted using `anonymous` memory, while microVMs
  resumed from snapshot load memory on-demand from the snapshot and copy-on-write
  to anonymous memory.

> - Resuming from a snapshot is optimized for speed. Taking a snapshot always involves

Author: I rephrased it, but I'm not sure if it sounds better now.

  synchronously writing dirty memory pages to the memory snapshot file. Taking a
  snapshot of a fresh microVM running off memory, results in
  full contents of guest memory being written to the snapshot.

> - The snapshot operations have minimal requirements/dependencies on when/how they

Author: Yep, kind of, but I agree it it's not clear at all, removed it.

  should be used.

> - The API calls exposing the snapshotting functionality have clear **Prerequisites** that

Author: :+1:

  describe the previously mentioned minimal requirements.
- The _memory file_ and _microVM state file_ are generated by Firecracker on snapshot
  creation and the disk contents are flushed to their backing files.
 
## Snapshot versioning

Firecracker snapshotting implementation offers support for microVM versioning
(`cross-version snapshots`) in the following contexts:
- saving snapshots at older versions,
- loading snapshots from older versions.

The value of `o` (maximum number of older versions that we can restore from / save
a snapshot to, from the current version) is not yet decided, but we plan on offering

> backwards compatibility for at least one year changes. 

Author: I added some more explanations to the whole chapter (`Snapshot versioning`), if it's too verbose I can simplify it back.

## Creating snapshots

> The supported and of interest operations in the snapshotting context are: `Pause`,

Author: Done.

`Resume`, `CreateSnapshot` and `LoadSnapshot`. `Pause`, `Resume` and `CreateSnapshot`
can be called only after booting the microVM, while `LoadSnapshot` is allowed only
before boot.

> For creating a snapshot, you have to pause the running microVM, which means pausing

Author: Done.

is consisted of the memory pages which have been dirtied since the last snapshot creation
or since the creation of the microVM, whichever of these events was the most recent.
All the other effects mentioned in the **Effects** section from `### Creating full snapshots`
apply here.
- _on failure_: no side-effects.

*Note*: To be able to do any incremental (diff) snapshots, you must set `track_dirty_pages`
field to `true` when configuring the `/machine-config` resource. Example:

```bash
curl --unix-socket /tmp/firecracker.socket -i  \
    -X PUT 'http://localhost/machine-config' \
    -H 'Accept: application/json'            \
    -H 'Content-Type: application/json'      \
    -d '{
            "vcpu_count": 2,
            "mem_size_mib": 1024,
            "ht_enabled": false,
            "track_dirty_pages": true
    }'
```

Enabling this support enables KVM dirty page tracking, so it comes at a cost; you should
only use it when needed.

Creating a snapshot will **not** influence state, will **not** stop or end the microVM,
it can be used as before, so you can resume the microVM if you still want to use it.

> Resuming the microVM can be done by sending the following API command:

Author: Hmm, I was thinking about that when I implemented the API calls. I don't think we should provide some kind of support for json configuration, wouldn't that json contain only a logger (optional I suppose) and the `LoadSnapshot` API body?

> Resuming the microVM can be done by sending the following API command:

Reviewer: It would definitely be a short one, but I think this would be pretty useful to one shot load  and resume. Let's discuss this offline since it is out of the scope of this PR.

```bash
curl --unix-socket /tmp/firecracker.socket -i \
    -X PATCH "http://localhost/vm" \
    -H "accept: application/json" \
    -H "Content-Type: application/json" \
    -d "{
            "state": "Resumed"
    }"
```

**Prerequisites**: You can resume the microVM any time after the microVM is booted.

>                    Successive calls of this request keep the microVM in the `Resumed`

Reviewer: This is not clear. The resume call does something only if the microvm is paused otherwise it is ignored.

Author: I rephrased it.

                   state.
**Effects**:
- _on success_: microVM is guaranteed to be `Resumed`.
- _on failure_: no side-effects.

## Loading snapshots

If you want to load a snapshot, you can do that only **before** the microVM is configured
(the only resource that can be configured prior is the Logger) by sending the following
API command:

```bash
curl --unix-socket /tmp/firecracker.socket -i \
    -X PUT "http://localhost/snapshot/load" \
    -H  "accept: application/json" \
    -H  "Content-Type: application/json" \
    -d "{
            \"snapshot_path\": \"/path/to/snapshot_file\",
	        \"mem_file_path\": \"/path/to/mem_file\",
            \"enable_diff_snapshots\": true
    }"
```

Details about the required and optional fields can be found in the
[swagger definition](../../src/api_server/swagger/firecracker.yaml).

> **Prerequisites**: Does **not** work with diff snapshots.

Reviewer: That is not a prerequisite.
```suggestion
**Prerequisites**: Caller needs to provide a full memory snapshot.
```

Author: True, I rephrased you suggestion a bit to keep it similar with the other prerequisites.

**Effects:**
- _on success_:
  - The complete microVM state will be loaded from snapshot into the current Firecracker
    process.

>   - It resets then the dirtied page bitmap and marks all pages clean (from an incremental

Reviewer: incremental or diff ? that is the question 😄

Author: I fail at consistency :)).

    snapshot point of view).

>   - The loaded microVM is now in the `Paused` state and needs to be resumed for it to run.

Reviewer: ```suggestion
  - The loaded microVM is now in the `Paused` state.
```

Author: I don't know why but it seems to me like an unfinished phrase. I kept it similar with the previous form, but if the second part is too obvious though, I can remove it.

>   - The file indicated by `mem_file_path` from which this microVM was resumed
>     **must** not change throughout the lifetime of the microVM. Since the guest in the
>     microVM will load pages from the file on demand, the Firecracker process should be
>     considered to have a shared-read-lock on the file. Any external writes to the file can
>     lead to microVM guest *Undefined Behavior*.

Reviewer: ```suggestion
  - The memory file pointed by `mem_file_path` **must** be considered immutable from Firecracker
  and host point of view. It backs the guest OS memory for read access through the page cache. 
  External modification to this file will corrupt the guest memory and leads to undefined behavior.
```

Author: Done.

  - The file indicated by `snapshot_path`, that is used to load from, is released and no
    longer used by this process.
  - If `enable_diff_snapshots` is set, then incremental snapshots can be taken afterwards.

> - _on failure_: The encountered error is reported, then the current Firecracker process

Reviewer: ```suggestion
- _on failure_: A specific error is reported and then the current Firecracker process
```

Author: Done.

# Firecracker Snapshotting

## What is microVM snapshotting?

MicroVM snapshotting is a mechanism through which a running microVM, including
the guest environment, can be serialized and saved to an external medium in the
form of a `snapshot`. This snapshot can be later used to restore a microVM
with its guest workload at that particular point in time.

## Snapshotting in Firecracker

> Firecracker snapshotting allows the creation of a microVM snapshot. This

Reviewer 2: `Firecracker snapshotting allows the creation of a microVM snapshot.` is a bit redundant, I think it can be removed.

snapshot can be later loaded in a possibly different Firecracker process and
the original guest workload is simply resumed. It is not guaranteed though that
the state of the network connections survives the process.

In order to make restoring possible, Firecracker snapshots save the full state
of the following resources:
- the guest memory,
- the emulated HW state (both KVM and Firecracker emulated HW),

> - the guest's storage disks' contents.

Reviewer 2: Firecracker doesn't actively do anything to save the disk separately. That's called out a few lines below; I'd remove mention of disk contents here.

The state of the components listed above is generated independently, which brings
flexibility to our snapshotting support. This means that taking a snapshot results
in multiple files that are composing the full microVM snapshot:
- the guest memory file,
- the microVM state file,
- zero or more disk files (depending on how many the guest had; these are
  **managed by the users**, which means they need to externally back up their
  block devices backing files).

This design allows sharing any read-only resources (such as read-only disks or
microVM state) between multiple microVMs in order to improve density.
When loading a snapshot, instead of loading at resume time the full contents from

> file to memory, Firecracker creates a `MAP_PRIVATE` mapping of the memory file,

Reviewer 2: A link to `mmap` reference would be useful here, as supporting explanation for what `MAP_PRIVATE` means

resulting in runtime on-demand loading of memory pages. Any subsequent memory writes
go to a copy-on-write anonymous memory mapping.
This has the advantage of very fast snapshot loading times, but comes with the cost
of having to keep the guest memory file around for the entire lifetime of the
resumed microVM.

## Firecracker Snapshotting characteristics

- Fresh Firecracker microVMs are booted using `anonymous` memory, while microVMs
  resumed from snapshot load memory on-demand from the snapshot and copy-on-write
  to anonymous memory.
- Resuming from a snapshot is optimized for speed. Taking a snapshot always involves
  synchronously writing dirty memory pages to the memory snapshot file. Taking a

>   snapshot of a fresh microVM running off non-file-backed anonymous memory, results in

Reviewer 2: `non-file-backed anonymous` is redundant

  full contents of guest memory being written to the snapshot.
- The snapshot operations have minimal requirements/dependencies on when/how they
  should be used.
- The API calls exposing the snapshotting functionality have clear **Prerequisites** that
  describe the previously mentioned minimal requirements.
- The _memory file_ and _microVM state file_ are generated by Firecracker on snapshot
  creation and the disk contents are flushed to their backing files.
 
## Snapshot versioning

Firecracker snapshotting implementation offers support for microVM versioning
(`cross-version snapshots`) in the following contexts:
- saving snapshots at older versions (also called `rollback`),
- loading snapshots from older versions (also called `live update`).
Live update refers to successfully loading any snapshot created by Firecracker version
`N` in a Firecracker version `N+o`.
Rollback means being able to create a snapshot with any version in `[N, N + o]` interval
while being in Firecracker version `N+o`.

The value of `o` (maximum number of older versions that we can restore from / save
a snapshot to, from the current version) is not yet decided, but we plan on offering

> backwards compatibility for at least one year changes. 

Reviewer 2: This sounds like a promise and we don't have anything decided yet. Please remove.

## Creating snapshots

The supported and of interest operations in the snapshotting context are: `Pause`,

> `Resume`, `CreateSnapshot` and `LoadSnapshot`. `Pause`, `Resume` and `CreateSnapshot`

Reviewer 2: I think it would be clearer if each had its own subchapter with a brief explanation, conditions, limitations and purpose

can be called only after booting the microVM, while `LoadSnapshot` is allowed only
before boot.

> For creating a snapshot, you have to pause the running microVM, which means pausing

Reviewer 2: Please use the same voice throughout the file. Either:
`you must do X`
or
`users must do X`
or
`X must be done`

I prefer option 2 but that's a weak personal preference, the strong preference being that the addressing is consistent

the microVM' vCPUs. If you change your mind and want to create a snapshot later,
you can resume the microVM and then pause it again when you decide to take that
snapshot.

For pausing the microVM you should send the following API command:

```bash
curl --unix-socket /tmp/firecracker.socket -i \
    -X PATCH "http://localhost/vm" \
    -H "accept: application/json" \
    -H "Content-Type: application/json" \
    -d "{
            "state": "Paused"
    }"
```

**Prerequisites**: You can pause the microVM any time after the microVM is booted.
                   Successive calls of this request keep the microVM in the `Paused`
                   state.
**Effects**:
- _on success_: microVM is guaranteed to be `Paused`.
- _on failure_: no side-effects.

> Now that the microVM is paused, you can create a snapshot, which can be either a `full`

Reviewer 2: Full vs diff snapshots are a big enough thing to deserve their own chapter, explaining their utility in various use cases, and how to use diffs.

Author: I made the separation a little bit clearer.

    }"
```

**Prerequisites**: It can be called only when the microVM is `Paused`.
**Effects**:
- _on success_:
The file indicated by `/path/to/snapshot_file` contains the devices' model state and
emulation state, same as when creating a full snapshot. The one indicated by
`/path/to/mem_file` contains this time a **diff copy** of the guest memory - the diff
is consisted of the memory pages which have been dirtied since the last snapshot creation
or since the creation of the microVM, whichever of these events was the most recent.
All the other effects mentioned in the **Effects** section from `### Creating full snapshots`
apply here.
- _on failure_: no side-effects.

*Note*: To be able to do any incremental (diff) snapshots, you must set `track_dirty_pages`
field to `true` when configuring the `/machine-config` resource. Example:

```bash
curl --unix-socket /tmp/firecracker.socket -i  \
    -X PUT 'http://localhost/machine-config' \
    -H 'Accept: application/json'            \
    -H 'Content-Type: application/json'      \
    -d '{
            "vcpu_count": 2,
            "mem_size_mib": 1024,
            "ht_enabled": false,
            "track_dirty_pages": true
    }'
```

> Enabling this support enables KVM dirty page tracking, so it comes at a cost; you should

Reviewer 2: The cost should at least be mentioned (CPU cycles spent by KVM accounting for dirtied pages).

    microVM will load pages from the file on demand, the Firecracker process should be
    considered to have a shared-read-lock on the file. Any external writes to the file can
    lead to microVM guest *Undefined Behavior*.
  - The file indicated by `snapshot_path`, that is used to load from, is released and no
    longer used by this process.
  - If `enable_diff_snapshots` is set, then incremental snapshots can be taken afterwards.
- _on failure_: The encountered error is reported, then the current Firecracker process
                is ended (as it might be in an invalid state).

Please, keep in mind that only if you set to true `enable_diff_snapshots`, when loading a
snapshot, or `track_dirty_pages`, when configuring the machine on a fresh microVM, you can
then create a `diff` snapshot. Also, `track_dirty_pages` is not saved when creating a
snapshot, so you need to explicitly set `enable_diff_snapshots` when sending `LoadSnapshot`
command if you want to be able to do diff snapshots from a loaded microVM.
Another thing that you should be aware of is the following: if a fresh microVM can create

> incremental snapshots, then if you create a **full** snapshot, the memory file contains

Reviewer 2: Please use `diff` snapshots instead of `incremental`

Author: I removed all `incremental` occurrences.

the whole guest memory, while if you create a **diff** one, that file contains only the
written pages.

## Snapshot Tools

> To enable users to benefit from incremental snapshotting, we intend to provide a tool that

Reviewer 2: This should contain a link to an issue where we're tracking it

Author: True.

(`cross-version snapshots`) in the following contexts:
- saving snapshots at older versions (being able to create a snapshot with any version
in `[N, N + o]` interval while being in Firecracker version `N+o`),
- loading snapshots from older versions (being able to load any snapshot created by
Firecracker version `N` in a Firecracker version `N+o`).

The design supports an unlimited number of versions, the value of `o` (maximum number
of older versions that we can restore from / save a snapshot to, from the current
version) will be defined later.

## Snapshot API

Firecracker exposes the following APIs for manipulating snapshots: `Pause`, `Resume`
and `CreateSnapshot` can be called only after booting the microVM, while `LoadSnapshot`
is allowed only before boot.

### Pausing the microVM

To create a snapshot, you have to firstly pause the running microVM and its vCPUs with
the following API command:

```bash
curl --unix-socket /tmp/firecracker.socket -i \
    -X PATCH "http://localhost/vm" \
    -H "accept: application/json" \
    -H "Content-Type: application/json" \
    -d "{
            "state": "Paused"
    }"
```

> **Prerequisites**: Pausing the microVM can be done any time after the microVM is booted.

Reviewer: There is not need to repeat the title. You can simply say: Microvm is booted.

Author: Done.

                   Successive calls of this request keep the microVM in the `Paused`
                   state.
**Effects**:
- _on success_: microVM is guaranteed to be `Paused`.
- _on failure_: no side-effects.

> *Note*: If you change your mind and want to create a snapshot later, you can resume the

Reviewer: I would remove this note. This can be rephrased better in the context of creating snapshots, highlighting the possible flows of boot -> pause -> save -> resume -> pause -> save, load->resume->pause->save->resume which work with either diff or full snapshots.

Author: Done, I updated with these flows at the end of the doc (after the `load` part).

        microVM and then pause it again when you decide to take that snapshot.

## Creating snapshots

Now that the microVM is paused, you can create a snapshot, which can be either a `full`
one or a `diff` one. Full snapshots always create a complete, resume-able snapshot of
the current microVM state and memory. Diff snapshots save the current microVM state
and the memory dirtied since the last snapshot (full or diff). Diff snapshots are not
resume-able, but can be merged into a full snapshot using external (provided) tooling.

### Creating full snapshots

For creating a full snapshot, you can use the following API command:

```bash
curl --unix-socket /tmp/firecracker.socket -i \
    -X PUT "http://localhost/snapshot/create" \
    -H  "accept: application/json" \
    -H  "Content-Type: application/json" \
    -d "{
            \"snapshot_type\": \"Full\",
            \"snapshot_path\": \"/path/to/snapshot_file\",
	        \"mem_file_path\": \"/path/to/mem_file\",
            \"version\": \"2\"
    }"
```

Details about the required and optional fields can be found in the
[swagger definition](../../src/api_server/swagger/firecracker.yaml).

> **Prerequisites**: It can be called only when the microVM is `Paused`.

Reviewer: this can be shorter: Microvm is Paused.

Author: Done.

**Effects**:
- _on success_:
  - The file indicated by `/path/to/snapshot_file` contains the devices' model state
    and emulation state. The one indicated by `/path/to/mem_file` contains a full copy
    of the guest memory.
  - The block devices contents are flushed to their backing files. At this point, these
    should be backed up externally by the user. The generated snapshot files are
    immediately available to be used (current process releases ownership).

>   - The snapshot creation resets then the dirtied page bitmap and marks all pages clean

Reviewer: This should happen only if diff snapshots are enabled when booting/restoring.

Author: Updated it.

    (from a diff snapshot point of view).

If a `version` is specified, the new snapshot is saved at that version, otherwise
it will be saved at the same version of the running Firecracker. The version is only
used for the microVM state file as it contains internal state structures for device
emulation, vCPUs and others that can change their format from a Firecracker version

> to another. Versioning is not supported for block devices and memory files. The separate

Reviewer: ```suggestion
to another. Versioning is not required for block and memory files. The separate
```

Author: Done.

block device file components of the snapshot have to be handled by the user.

- _on failure_: no side-effects.

### Creating diff snapshots

For creating a diff snapshot, you should use the same API command, but with
`snapshot_type` field set to `Diff`.

> *Note*: If omitted from the request body, `snapshot_type` default value is `Full`.

Reviewer: ```suggestion
*Note*: If not specified, `snapshot_type` defaults to`Full`.
```

Author: `defaults` in this context is a phrasal verb and I don't particularly like using it in a doc; replaced it with `is by default`.

```bash
curl --unix-socket /tmp/firecracker.socket -i \
    -X PUT "http://localhost/snapshot/create" \
    -H  "accept: application/json" \
    -H  "Content-Type: application/json" \
    -d "{
            \"snapshot_type\": \"Diff\",
            \"snapshot_path\": \"/path/to/snapshot_file\",
	        \"mem_file_path\": \"/path/to/mem_file\",
            \"version\": \"2\"
    }"
```

**Prerequisites**: It can be called only when the microVM is `Paused`.
**Effects**:
- _on success_:
The file indicated by `/path/to/snapshot_file` contains the devices' model state and
emulation state, same as when creating a full snapshot. The one indicated by
`/path/to/mem_file` contains this time a **diff copy** of the guest memory - the diff
is consisted of the memory pages which have been dirtied since the last snapshot creation
or since the creation of the microVM, whichever of these events was the most recent.
All the other effects mentioned in the **Effects** section from `## Creating full snapshots`
apply here.
- _on failure_: no side-effects.

To be able to do any diff snapshots, `track_dirty_pages` field should be set to `true`

> when configuring the `/machine-config` resource. Example:

Reviewer: This is a prerequisite.

Reviewer: Also you would need to detail both the boot/load setting of track_dirty_pages.

Author: Done.

```bash
curl --unix-socket /tmp/firecracker.socket -i  \
    -X PUT 'http://localhost/machine-config' \
    -H 'Accept: application/json'            \
    -H 'Content-Type: application/json'      \
    -d '{
            "vcpu_count": 2,
            "mem_size_mib": 1024,
            "ht_enabled": false,
            "track_dirty_pages": true
    }'
```

*Note*: Enabling this support enables KVM dirty page tracking, so it comes at a cost
(which consists of CPU cycles spent by KVM accounting for dirtied pages); it should only
be used when needed.

Creating a snapshot will **not** influence state, will **not** stop or end the microVM,
it can be used as before, so the microVM can be resumed if you still want to use it.

### Resuming the microVM

You can resume the microVM by sending the following API command:

```bash
curl --unix-socket /tmp/firecracker.socket -i \
    -X PATCH "http://localhost/vm" \
    -H "accept: application/json" \
    -H "Content-Type: application/json" \
    -d "{
            "state": "Resumed"
    }"
```

> **Prerequisites**: The microVM can be resumed any time after the microVM is booted.

Reviewer: Microvm is `Paused`.

Author: Done.

                   Successive calls of this request are ignored (microVM remains
                   in the running state).
**Effects**:
- _on success_: microVM is guaranteed to be `Resumed`.
- _on failure_: no side-effects.

## Loading snapshots

If you want to load a snapshot, you can do that only **before** the microVM is configured
(the only resource that can be configured prior is the Logger) by sending the following
API command:

```bash
curl --unix-socket /tmp/firecracker.socket -i \
    -X PUT "http://localhost/snapshot/load" \
    -H  "accept: application/json" \
    -H  "Content-Type: application/json" \
    -d "{
            \"snapshot_path\": \"/path/to/snapshot_file\",
	        \"mem_file_path\": \"/path/to/mem_file\",
            \"enable_diff_snapshots\": true
    }"
```

Details about the required and optional fields can be found in the
[swagger definition](../../src/api_server/swagger/firecracker.yaml).

> **Prerequisites**: A full memory snapshot **must** be provided.

Reviewer: ```suggestion
**Prerequisites**: A full memory snapshot and a vmstate file **must** be provided.
```

    Firecracker and host point of view. It backs the guest OS memory for read access
    through the page cache. External modification to this file corrupts the guest
    memory and leads to undefined behavior.
  - The file indicated by `snapshot_path`, that is used to load from, is released and no
    longer used by this process.
  - If `enable_diff_snapshots` is set, then diff snapshots can be taken afterwards.
- _on failure_: A specific error is reported and then the current Firecracker process
                is ended (as it might be in an invalid state).

*Notes*:
Please, keep in mind that only by setting to true `enable_diff_snapshots`, when loading a
snapshot, or `track_dirty_pages`, when configuring the machine on a fresh microVM, you can
then create a `diff` snapshot. Also, `track_dirty_pages` is not saved when creating a
snapshot, so you need to explicitly set `enable_diff_snapshots` when sending `LoadSnapshot`
command if you want to be able to do diff snapshots from a loaded microVM.
Another thing that you should be aware of is the following: if a fresh microVM can create
diff snapshots, then if you create a **full** snapshot, the memory file contains

> the whole guest memory, while if you create a **diff** one, that file contains only the

Reviewer: ```suggestion
the whole guest memory, while if you create a **diff** one, that file is sparse and only contains the guest dirtied pages
```

Author: Done.

# Firecracker Snapshotting

## What is microVM snapshotting?

MicroVM snapshotting is a mechanism through which a running microVM and its
resources can be serialized and saved to an external medium in the form of a
`snapshot`. This snapshot can be later used to restore a microVM with its
guest workload at that particular point in time.

## Snapshotting in Firecracker

> A Firecracker microVM snapshot can be used for loading it later in a possibly

Author: Updated it.

```

Enabling this support enables KVM dirty page tracking, so it comes at a cost
(which consists of CPU cycles spent by KVM accounting for dirtied pages); it should only
be used when needed.

Creating a snapshot will **not** influence state, will **not** stop or end the microVM,
it can be used as before, so the microVM can be resumed if you still want to use it.

### Resuming the microVM

You can resume the microVM by sending the following API command:

```bash
curl --unix-socket /tmp/firecracker.socket -i \
    -X PATCH "http://localhost/vm" \
    -H "accept: application/json" \
    -H "Content-Type: application/json" \
    -d "{
            "state": "Resumed"
    }"
```

**Prerequisites**: The microVM is `Paused`.
                   Successive calls of this request are ignored (microVM remains
                   in the running state).
**Effects**:
- _on success_: microVM is guaranteed to be `Resumed`.
- _on failure_: no side-effects.

## Loading snapshots

If you want to load a snapshot, you can do that only **before** the microVM is configured
(the only resources that can be configured prior are the Logger and the Metrics systems)
by sending the following API command:

```bash

> curl --unix-socket /tmp/firecracker.socket -i \
>     -X PUT "http://localhost/snapshot/load" \
>     -H  "accept: application/json" \
>     -H  "Content-Type: application/json" \
>     -d "{
>             \"snapshot_path\": \"/path/to/snapshot_file\",
> 	        \"mem_file_path\": \"/path/to/mem_file\",
>             \"enable_diff_snapshots\": true
>     }"

Reviewer 3: ```suggestion
curl --unix-socket /tmp/firecracker.socket -i \
    -X PUT "http://localhost/snapshot/load" \
    -H  "accept: application/json" \
    -H  "Content-Type: application/json" \
    -d "{
           \"snapshot_path\": \"/path/to/snapshot_file\",
	     \"mem_file_path\": \"/path/to/mem_file\",
           \"enable_diff_snapshots\": true
    }"
```

Author: looks like this happened many times :(

```

Details about the required and optional fields can be found in the
[swagger definition](../../src/api_server/swagger/firecracker.yaml).

**Prerequisites**: A full memory snapshot and a microVM state file **must** be provided.
**Effects:**
- _on success_:
  - The complete microVM state is loaded from snapshot into the current Firecracker
    process.

>   - It resets then the dirtied page bitmap and marks all pages clean (from a diff

Reviewer 3: ```suggestion
  - It then resets the dirtied page bitmap and marks all pages clean (from a diff
```

Author: done.

emulation state, same as when creating a full snapshot. The one indicated by
`/path/to/mem_file` contains this time a **diff copy** of the guest memory - the diff
consists of the memory pages which have been dirtied since the last snapshot creation
or since the creation of the microVM, whichever of these events was the most recent.
All the other effects mentioned in the **Effects** section from `## Creating full snapshots`
apply here.
- _on failure_: no side-effects.

*Note*: This is an example of an API command that enables dirty page tracking:

```bash
curl --unix-socket /tmp/firecracker.socket -i  \
    -X PUT 'http://localhost/machine-config' \
    -H 'Accept: application/json'            \
    -H 'Content-Type: application/json'      \
    -d '{
            "vcpu_count": 2,
            "mem_size_mib": 1024,
            "ht_enabled": false,
            "track_dirty_pages": true
    }'
```

Enabling this support enables KVM dirty page tracking, so it comes at a cost
(which consists of CPU cycles spent by KVM accounting for dirtied pages); it should only
be used when needed.

Creating a snapshot will **not** influence state, will **not** stop or end the microVM,

> it can be used as before, so the microVM can be resumed if you still want to use it.

Reviewer 4: nit: I would stress the disks story even further here:
```
Users should make sure to copy disk backing files at this point if they plan to continue using the current microVM.
```

Author: Done.

    -X PATCH "http://localhost/vm" \
    -H "accept: application/json" \
    -H "Content-Type: application/json" \
    -d "{
            "state": "Resumed"
    }"
```

**Prerequisites**: The microVM is `Paused`.
                   Successive calls of this request are ignored (microVM remains
                   in the running state).
**Effects**:
- _on success_: microVM is guaranteed to be `Resumed`.
- _on failure_: no side-effects.

## Loading snapshots

If you want to load a snapshot, you can do that only **before** the microVM is configured
(the only resources that can be configured prior are the Logger and the Metrics systems)
by sending the following API command:

```bash
curl --unix-socket /tmp/firecracker.socket -i \
    -X PUT "http://localhost/snapshot/load" \
    -H  "accept: application/json" \
    -H  "Content-Type: application/json" \
    -d "{
            \"snapshot_path\": \"/path/to/snapshot_file\",
            \"mem_file_path\": \"/path/to/mem_file\",
            \"enable_diff_snapshots\": true
    }"
```

Details about the required and optional fields can be found in the
[swagger definition](../../src/api_server/swagger/firecracker.yaml).

> **Prerequisites**: A full memory snapshot and a microVM state file **must** be provided.

Reviewer 4: What about adding disks, net ifaces, vsock backend management as a prerequisite as well?
Smth along the lines:

> Users must make sure _disk backing files, network interfaces backing TAPs and/or Vsock backend socket_ are set up and accessible to the new firecracker process where the microVM is resumed. These host-resources need to be accessible at the same relative paths to the new firecracker process as they were to the original firecracker process.

Author: Sounds good.

  block devices backing files).

The design allows sharing of memory pages and read only disks between multiple
microVMs. When loading a snapshot, instead of loading at resume time the full
contents from file to memory, Firecracker creates a
[MAP_PRIVATE mapping](http://man7.org/linux/man-pages/man2/mmap.2.html) of the
memory file, resulting in runtime on-demand loading of memory pages. Any subsequent
memory writes go to a copy-on-write anonymous memory mapping.
This has the advantage of very fast snapshot loading times, but comes with the cost
of having to keep the guest memory file around for the entire lifetime of the
resumed microVM.

## Firecracker Snapshotting characteristics

- Fresh Firecracker microVMs are booted using `anonymous` memory, while microVMs
  resumed from snapshot load memory on-demand from the snapshot and copy-on-write
  to anonymous memory.
- Resuming from a snapshot is optimized for speed, while taking a snapshot involves
  some extra CPU cycles for synchronously writing dirty memory pages to the memory
  snapshot file. Taking a snapshot of a fresh microVM, on which dirty pages tracking

>   is not enabled, results in full contents of guest memory being written to the

Reviewer 3: nit:
```suggestion
  is not enabled, results in the full contents of guest memory being written to the
```

  snapshot.
- The _memory file_ and _microVM state file_ are generated by Firecracker on snapshot
  creation, and the disk contents are flushed to their backing files.  
- The API calls exposing the snapshotting functionality have clear **Prerequisites**
  that describe the requirements on when/how they should be used.
 
## Snapshot versioning

Firecracker snapshotting implementation offers support for microVM versioning
(`cross-version snapshots`) in the following contexts:
- saving snapshots at older versions (being able to create a snapshot with any version
in `[N, N + o]` interval while being in Firecracker version `N+o`),
- loading snapshots from older versions (being able to load any snapshot created by

> Firecracker version `N` in a Firecracker version `N+o`).

Reviewer 3: nit:
```suggestion
Firecracker version in the `[N, N + o]` interval in Firecracker version `N+o`).
```

The design supports an unlimited number of versions, the value of `o` (maximum number
of older versions that we can restore from / save a snapshot to, from the current
version) will be defined later.

## Snapshot API

Firecracker exposes the following APIs for manipulating snapshots: `Pause`, `Resume`
and `CreateSnapshot` can be called only after booting the microVM, while `LoadSnapshot`
is allowed only before boot.

### Pausing the microVM

> To create a snapshot, you have to firstly pause the running microVM and its vCPUs with

Reviewer 3: very subjective nit:
```suggestion
To create a snapshot, first you have to pause the running microVM and its vCPUs with
```

the following API command:

```bash

> curl --unix-socket /tmp/firecracker.socket -i \

Reviewer 5: While using `curl`, the outer quotes must be singular and the inner ones must be double.

    -X PATCH "http://localhost/vm" \
    -H "accept: application/json" \
    -H "Content-Type: application/json" \
    -d "{
            "state": "Paused"
    }"
```

**Prerequisites**: The microVM is booted.
                   Successive calls of this request keep the microVM in the `Paused`
                   state.
**Effects**:
- _on success_: microVM is guaranteed to be `Paused`.
- _on failure_: no side-effects.

## Creating snapshots

Now that the microVM is paused, you can create a snapshot, which can be either a `full`
one or a `diff` one. Full snapshots always create a complete, resume-able snapshot of
the current microVM state and memory. Diff snapshots save the current microVM state
and the memory dirtied since the last snapshot (full or diff). Diff snapshots are not
resume-able, but can be merged into a full snapshot using external (provided) tooling.

### Creating full snapshots

For creating a full snapshot, you can use the following API command:

```bash

> curl --unix-socket /tmp/firecracker.socket -i \

Reviewer 5: Same here but there are also slashes that need to be removed.

Author: Many thanks for the heads-up! The document should be more consistent now.
I also added a note that specifies the fact that if the files indicated by the snapshot and memory paths don't exist, then they will be created.

    -X PUT "http://localhost/snapshot/create" \
    -H  "accept: application/json" \
    -H  "Content-Type: application/json" \
    -d "{
            \"snapshot_type\": \"Full\",
            \"snapshot_path\": \"/path/to/snapshot_file\",
            \"mem_file_path\": \"/path/to/mem_file\",

>             \"version\": \"2\"

Reviewer 3: ```suggestion
            \"version\": \"0.23.0\"
```

    of the guest memory.
  - The block devices contents are flushed to their backing files. At this point, these
    should be backed up externally by the user. The generated snapshot files are
    immediately available to be used (current process releases ownership).
  - If diff snapshots were enabled, the snapshot creation resets then the dirtied page
    bitmap and marks all pages clean (from a diff snapshot point of view).

If a `version` is specified, the new snapshot is saved at that version, otherwise
it will be saved at the same version of the running Firecracker. The version is only
used for the microVM state file as it contains internal state structures for device
emulation, vCPUs and others that can change their format from a Firecracker version
to another. Versioning is not required for the block and memory files. The separate
block device file components of the snapshot have to be handled by the user.

- _on failure_: no side-effects.

### Creating diff snapshots

For creating a diff snapshot, you should use the same API command, but with
`snapshot_type` field set to `Diff`.
*Note*: If not specified, `snapshot_type` is by default `Full`.

```bash

> curl --unix-socket /tmp/firecracker.socket -i \

Reviewer 5: same here

    -X PUT "http://localhost/snapshot/create" \
    -H  "accept: application/json" \
    -H  "Content-Type: application/json" \
    -d "{
            \"snapshot_type\": \"Diff\",
            \"snapshot_path\": \"/path/to/snapshot_file\",
            \"mem_file_path\": \"/path/to/mem_file\",

>             \"version\": \"2\"

Reviewer 3: ```suggestion
            \"version\": \"0.23.0\"
```

    }"
```

**Prerequisites**: The microVM is `Paused`.
                   On a fresh microVM, `track_dirty_pages` field should be set to `true`,
                   when configuring the `/machine-config` resource, while on a snapshot
                   loaded microVM, `enable_diff_snapshots` from `PUT /snapshot/load`
                   request body, should be set.
                   
**Effects**:
- _on success_:
The file indicated by `/path/to/snapshot_file` contains the devices' model state and
emulation state, same as when creating a full snapshot. The one indicated by
`/path/to/mem_file` contains this time a **diff copy** of the guest memory - the diff
consists of the memory pages which have been dirtied since the last snapshot creation
or since the creation of the microVM, whichever of these events was the most recent.
All the other effects mentioned in the **Effects** section from `## Creating full snapshots`
apply here.
- _on failure_: no side-effects.

*Note*: This is an example of an API command that enables dirty page tracking:

```bash

> curl --unix-socket /tmp/firecracker.socket -i  \

Reviewer 5: this one has the correct format of the quotes

    -X PUT 'http://localhost/machine-config' \
    -H 'Accept: application/json'            \
    -H 'Content-Type: application/json'      \
    -d '{
            "vcpu_count": 2,
            "mem_size_mib": 1024,
            "ht_enabled": false,
            "track_dirty_pages": true
    }'
```

Enabling this support enables KVM dirty page tracking, so it comes at a cost
(which consists of CPU cycles spent by KVM accounting for dirtied pages); it should only
be used when needed.

Creating a snapshot will **not** influence state, will **not** stop or end the microVM,
it can be used as before, so the microVM can be resumed if you still want to use it.
At this point, in case you plan to continue using the current microVM, you should make
sure to also copy the disk backing files.

### Resuming the microVM

You can resume the microVM by sending the following API command:

```bash

> curl --unix-socket /tmp/firecracker.socket -i \

Reviewer 5: same problem here

    -X PATCH "http://localhost/vm" \
    -H "accept: application/json" \
    -H "Content-Type: application/json" \
    -d "{
            "state": "Resumed"
    }"
```

**Prerequisites**: The microVM is `Paused`.
                   Successive calls of this request are ignored (microVM remains
                   in the running state).
**Effects**:
- _on success_: microVM is guaranteed to be `Resumed`.
- _on failure_: no side-effects.

## Loading snapshots

If you want to load a snapshot, you can do that only **before** the microVM is configured
(the only resources that can be configured prior are the Logger and the Metrics systems)
by sending the following API command:

```bash

> curl --unix-socket /tmp/firecracker.socket -i \

Reviewer 5: and here, also the slashes
